zkevm-circuits/src/anchor_tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,459 @@
+    util::{split_u256, Expr},
+};
+use halo2_proofs::{
+    circuit::{Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, SecondPhase, Selector},
+    poly::Rotation,
+};
+use once_cell::sync::Lazy;
+use std::str::FromStr;
+
+const MAX_DEGREE: usize = 9;
+const BYTE_POW_BASE: u64 = 1 << 8;
+
+pub(crate) static GOLDEN_TOUCH_ADDRESS: Lazy<Address> =
+    Lazy::new(|| address!("0x0000777735367b36bC9B61C50022d9D0700dB4Ec"));
+
+// 0x92954368afd3caa1f3ce3ead0069c1af414054aefe1ef9aeacc1bf426222ce38
+pub(crate) static GOLDEN_TOUCH_PRIVATEKEY: Lazy<Word> =
+    Lazy::new(|| word!("0x92954368afd3caa1f3ce3ead0069c1af414054aefe1ef9aeacc1bf426222ce38"));
+
+pub(crate) static GOLDEN_TOUCH_WALLET: Lazy<LocalWallet> = Lazy::new(|| {
+    LocalWallet::from_str("0x92954368afd3caa1f3ce3ead0069c1af414054aefe1ef9aeacc1bf426222ce38")
+        .unwrap()
+});
+
+// 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798
+pub(crate) static GX1: Lazy<Word> =
+    Lazy::new(|| word!("0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798"));
+static GX1_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX1));
+
+// 0xc6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5
+pub(crate) static GX2: Lazy<Word> =
+    Lazy::new(|| word!("0xc6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"));
+static GX2_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX2));
+
+// 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141
+pub(crate) static N: Lazy<Word> =
+    Lazy::new(|| word!("0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141"));
+static N_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&N));
+
+// private key 0x92954368afd3caa1f3ce3ead0069c1af414054aefe1ef9aeacc1bf426222ce38
+// GX1 * PRIVATEKEY(mod N) = 0x4341adf5a780b4a87939938fd7a032f6e6664c7da553c121d3b4947429639122
+pub(crate) static GX1_MUL_PRIVATEKEY: Lazy<Word> =
+    Lazy::new(|| word!("0x4341adf5a780b4a87939938fd7a032f6e6664c7da553c121d3b4947429639122"));
+static GX1_MUL_PRIVATEKEY_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX1_MUL_PRIVATEKEY));
+static GX1_MUL_PRIVATEKEY_LIMB64: Lazy<[U256; 4]> =
+    Lazy::new(|| split_u256_limb64(&GX1_MUL_PRIVATEKEY));
+
+// GX2 * PRIVATEKEY(mod N) = 0x4a43b192ca74cab200d6c086df90fb729abca9e52d38b8fa0beb4eafe70956de
+static GX2_MUL_PRIVATEKEY: Lazy<Word> =
+    Lazy::new(|| word!("0x4a43b192ca74cab200d6c086df90fb729abca9e52d38b8fa0beb4eafe70956de"));
+static GX2_MUL_PRIVATEKEY_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX2_MUL_PRIVATEKEY));
+
+// In mul_add chip, we have a * b + c == d
+// => a == GX1_MUL_PRIVATEKEY
+// => b == 1
+// => c == msg_hash
+// => d == N
+//
+// # The circuit layout
+// - msg_hash (c)
+// - SigR
+
+#[derive(Debug, Clone)]
+pub(crate) struct SignVerifyConfig<F: Field> {
+    tx_table: TxTable,
+
+    q_sign_start: Selector,
+    q_sign_step: Selector,
+    q_sign_end: Selector,
```

Comment:
similar to https://github.com/taikoxyz/zkevm-circuits/pull/105#discussion_r1235182175